File: apps/mobile-native/ios/NexaraApp/UI/Core/ChecklistVehiculoRules.swift
import Foundation

/// Reglas puras del check list de entrega y recepción de vehículos utilitarios.
///
/// Espejo de `apps/api/src/vehicles/checklist-entrega.ts` (la fuente de verdad)
/// y del `ChecklistVehiculoRules.kt` de Android: los mismos siete slots, las
/// mismas etiquetas cortas, los mismos niveles de gasolina y —palabra por
/// palabra— los mismos mensajes de error, para que la app no deje enviar algo
/// que el API va a rechazar con un 400.
///
/// Aquí no entra SwiftUI ni UIKit: solo Foundation, para que la regla se pueda
/// leer y comparar contra el TypeScript de un vistazo.

// MARK: - Slots

/// Las siete fotos del check list. El `rawValue` es el nombre del campo del
/// multipart y la clave dentro del JSON de `meta`.
enum SlotChecklist: String, CaseIterable, Identifiable, Hashable {
    // Las cuatro caras del 360°.
    case frontal
    case trasera
    case lateralIzq = "lateral-izq"
    case lateralDer = "lateral-der"
    // Habitáculo y cajuela.
    case interiorDelantera = "interior-delantera"
    case interiorTrasera = "interior-trasera"
    // Odómetro y aguja de gasolina en la misma foto.
    case tablero

    var id: String { rawValue }

    /// Nombre del campo del multipart (`SLOTS_CHECKLIST`).
    var field: String { rawValue }

    /// Etiqueta corta (`ETIQUETA_SLOT`). Sin prosa: cabe en un botón.
    var label: String {
        switch self {
        case .frontal: return "Frente"
        case .trasera: return "Trasera"
        case .lateralIzq: return "Lateral izq."
        case .lateralDer: return "Lateral der."
        case .interiorDelantera: return "Interior frente"
        case .interiorTrasera: return "Interior atrás"
        case .tablero: return "Tablero"
        }
    }

    /// Icono del paso en la cuadrícula.
    var systemImage: String {
        switch self {
        case .frontal: return "car.front.waves.up"
        case .trasera: return "car.rear"
        case .lateralIzq: return "car.side"
        case .lateralDer: return "car.side"
        case .interiorDelantera: return "carseat.left"
        case .interiorTrasera: return "carseat.right"
        case .tablero: return "gauge.with.dots.needle.bottom.50percent"
        }
    }
}

// MARK: - Combustible

/// Selector rápido junto a la foto del tablero (`NIVELES_COMBUSTIBLE`). Nadie
/// lee una aguja en porcentajes: se lee E, ¼, ½, ¾, F. El porcentaje es lo que
/// guarda el API.
enum NivelCombustible: String, CaseIterable, Identifiable, Hashable {
    case vacio = "E"
    case unCuarto = "1/4"
    case medio = "1/2"
    case tresCuartos = "3/4"
    case lleno = "F"

    var id: String { rawValue }

    /// Lo que viaja en el campo `combustible` del multipart.
    var apiValue: String { rawValue }

    /// Lo que se ve en el selector segmentado.
    var label: String {
        switch self {
        case .vacio: return "E"
        case .unCuarto: return "¼"
        case .medio: return "½"
        case .tresCuartos: return "¾"
        case .lleno: return "F"
        }
    }

    var pct: Int {
        switch self {
        case .vacio: return 0
        case .unCuarto: return 25
        case .medio: return 50
        case .tresCuartos: return 75
        case .lleno: return 100
        }
    }

    /// 40 → ½ (el más cercano). `nivelDesdeCombustiblePct` del API, para pintar
    /// de vuelta el nivel con el que salió el vehículo.
    static func from(pct: Int?) -> NivelCombustible? {
        guard let pct else { return nil }
        return allCases.min { abs($0.pct - pct) < abs($1.pct - pct) }
    }
}

// MARK: - Validación

/// Un error del check list: el campo que lo provoca y lo que se le dice a quien
/// está parado junto al coche (`ErrorChecklist` del API).
struct ErrorChecklist: Identifiable, Hashable {
    let campo: String
    let mensaje: String

    var id: String { "\(campo)|\(mensaje)" }
}

enum ChecklistVehiculoRules {
    /// Los siete slots en el orden en que se piden en pantalla.
    static let slots: [SlotChecklist] = SlotChecklist.allCases

    /// Paso 1 del flujo: las cuatro caras del 360°.
    static let exterior: [SlotChecklist] = [.frontal, .trasera, .lateralIzq, .lateralDer]

    /// Paso 1 del flujo: habitáculo y cajuela.
    static let interior: [SlotChecklist] = [.interiorDelantera, .interiorTrasera]

    /// Paso 2: el tablero va solo, con el kilometraje y la gasolina.
    static let tablero: SlotChecklist = .tablero

    /// Margen para aceptar una captura como «en vivo» (`VENTANA_CAPTURA_HORAS`).
    static let ventanaCapturaHoras: Double = 12

    /// Los slots que todavía no tienen foto.
    static func faltantes(capturas: [SlotChecklist: Date]) -> [SlotChecklist] {
        slots.filter { capturas[$0] == nil }
    }

    /// Todos los errores juntos, en el mismo orden que `revisarChecklist`: que
    /// nadie descubra el segundo error después de arreglar el primero.
    /// Vacío == se puede enviar.
    ///
    /// - Parameters:
    ///   - capturas: slot → instante en que la cámara tomó la foto
    ///     (`CapturedGeoPhoto.capturedAt`). Sin entrada == sin foto.
    ///   - odometroKm: lo tecleado en el paso de kilometraje; `nil` si está vacío.
    ///   - combustible: nivel marcado en el selector; `nil` si no se ha tocado.
    ///   - odometroInicio: en la devolución, el kilometraje con el que salió.
    ///   - ahora: inyectable para no depender del reloj real.
    static func validar(
        capturas: [SlotChecklist: Date],
        odometroKm: Int?,
        combustible: NivelCombustible?,
        odometroInicio: Int? = nil,
        ahora: Date = Date()
    ) -> [ErrorChecklist] {
        var errores: [ErrorChecklist] = []

        for slot in slots {
            guard let capturedAt = capturas[slot] else {
                errores.append(ErrorChecklist(campo: slot.field, mensaje: "Falta la foto: \(slot.label)"))
                continue
            }
            let edadHoras = ahora.timeIntervalSince(capturedAt) / 3600
            if edadHoras > ventanaCapturaHoras {
                errores.append(ErrorChecklist(
                    campo: slot.field,
                    mensaje: "\(slot.label): la foto no es de ahora, vuelve a tomarla"
                ))
            }
        }

        if let odometroKm {
            if odometroKm < 0 {
                errores.append(ErrorChecklist(campo: "odometroKm", mensaje: "Captura el kilometraje del tablero"))
            } else if let odometroInicio, odometroKm < odometroInicio {
                errores.append(ErrorChecklist(
                    campo: "odometroKm",
                    mensaje: "El kilometraje final no puede ser menor al inicial (\(odometroInicio) km)"
                ))
            }
        } else {
            errores.append(ErrorChecklist(campo: "odometroKm", mensaje: "Captura el kilometraje del tablero"))
        }

        if combustible == nil {
            errores.append(ErrorChecklist(
                campo: "combustible",
                mensaje: "Marca el nivel de gasolina (E, ¼, ½, ¾, F)"
            ))
        }

        return errores
    }

    /// Un solo renglón con todo lo que falta (`mensajeErrores` del API).
    static func mensaje(_ errores: [ErrorChecklist]) -> String {
        errores.map(\.mensaje).joined(separator: ". ")
    }
}

File: apps/mobile-native/android/app/src/main/java/mx/nexara/mobile/nativeapp/ui/console/vehiculos/ChecklistVehiculoRules.kt
package mx.nexara.mobile.nativeapp.ui.console.vehiculos

/**
 * Reglas del checklist de entrega/devolución de vehículo — Kotlin puro, sin
 * Android: es lo que se prueba en `ChecklistVehiculoRulesTest` contra
 * `apps/api/src/vehicles/checklist-entrega.spec.ts`.
 *
 * Siete fotos obligatorias, siempre con cámara en vivo. El servidor rechaza
 * cualquier foto sin `capturedAt` (así niega las subidas de galería), así que
 * aquí una foto sin hora cuenta como faltante y ni siquiera se manda.
 */
object ChecklistVehiculoRules {

    /** Dónde cae cada foto dentro del flujo por pasos. */
    enum class Grupo { EXTERIOR, INTERIOR, TABLERO }

    /** [id] es el nombre de la parte multipart; [etiqueta] lo que lee la persona. */
    data class Slot(val id: String, val etiqueta: String, val grupo: Grupo)

    const val FRONTAL = "frontal"
    const val TRASERA = "trasera"
    const val LATERAL_IZQ = "lateral-izq"
    const val LATERAL_DER = "lateral-der"
    const val INTERIOR_DELANTERA = "interior-delantera"
    const val INTERIOR_TRASERA = "interior-trasera"
    const val TABLERO = "tablero"

    /** Las 7 casillas, en el orden en que se piden. */
    val SLOTS: List<Slot> = listOf(
        Slot(FRONTAL, "Frente", Grupo.EXTERIOR),
        Slot(TRASERA, "Trasera", Grupo.EXTERIOR),
        Slot(LATERAL_IZQ, "Lateral izq.", Grupo.EXTERIOR),
        Slot(LATERAL_DER, "Lateral der.", Grupo.EXTERIOR),
        Slot(INTERIOR_DELANTERA, "Interior frente", Grupo.INTERIOR),
        Slot(INTERIOR_TRASERA, "Interior atrás", Grupo.INTERIOR),
        Slot(TABLERO, "Tablero", Grupo.TABLERO),
    )

    fun slots(grupo: Grupo): List<Slot> = SLOTS.filter { it.grupo == grupo }

    /** Paso 1 del flujo: 4 exterior + 2 interior. */
    val SLOTS_FOTOS: List<Slot> = SLOTS.filter { it.grupo != Grupo.TABLERO }

    fun etiqueta(id: String): String = SLOTS.firstOrNull { it.id == id }?.etiqueta ?: id

    // ── Combustible ─────────────────────────────────────────────────────────

    /** Selector rápido: E · ¼ · ½ · ¾ · F (el valor que viaja es el de la izquierda). */
    val NIVELES: List<String> = listOf("E", "1/4", "1/2", "3/4", "F")

    /** Cómo se dibuja cada nivel en el selector. */
    fun nivelCorto(nivel: String): String = when (nivel) {
        "1/4" -> "¼"
        "1/2" -> "½"
        "3/4" -> "¾"
        else -> nivel
    }

    /**
     * Porcentaje de un nivel. Acepta las cinco marcas del tanque o un número
     * 0–100 (el API admite ambas formas); cualquier otra cosa es `null`.
     */
    fun combustiblePct(nivel: String?): Int? {
        val v = nivel?.trim().orEmpty()
        if (v.isEmpty()) return null
        return when (v.uppercase()) {
            "E" -> 0
            "1/4" -> 25
            "1/2" -> 50
            "3/4" -> 75
            "F" -> 100
            else -> v.toIntOrNull()?.takeIf { it in 0..100 }
        }
    }

    // ── Validación ──────────────────────────────────────────────────────────

    /** Lo que el flujo guarda de cada foto; viaja al API dentro de `meta`. */
    data class SlotMeta(
        val capturedAt: String? = null,
        val lat: Double? = null,
        val lng: Double? = null,
    ) {
        /** Sin hora no sirve: el servidor la rechaza igual que una de galería. */
        val valida: Boolean get() = !capturedAt.isNullOrBlank()
    }

    data class Validacion(
        /** Etiquetas de las casillas sin foto (o con foto sin hora). */
        val faltantes: List<String> = emptyList(),
        /** Km y combustible: lo que hay que corregir, ya en español. */
        val errores: List<String> = emptyList(),
    ) {
        val ok: Boolean get() = faltantes.isEmpty() && errores.isEmpty()

        /** Una sola línea para el snackbar; `null` si todo está bien. */
        val mensaje: String?
            get() {
                if (ok) return null
                val partes = buildList {
                    if (faltantes.isNotEmpty()) add("Faltan fotos: ${faltantes.joinToString(", ")}")
                    addAll(errores)
                }
                return partes.joinToString(". ")
            }
    }

    /**
     * @param metas lo capturado hasta ahora, por id de casilla.
     * @param odometroKm kilometraje tecleado (`null` si está en blanco o no es número).
     * @param combustible nivel elegido (`E`, `1/4`, `1/2`, `3/4`, `F` o 0–100).
     * @param kmInicio en devolución, el km con el que salió: el final no puede ser menor.
     */
    fun validar(
        metas: Map<String, SlotMeta>,
        odometroKm: Int?,
        combustible: String?,
        kmInicio: Int? = null,
    ): Validacion {
        val faltantes = SLOTS.filter { metas[it.id]?.valida != true }.map { it.etiqueta }
        val errores = buildList {
            when {
                odometroKm == null -> add("Escribe el kilometraje")
                odometroKm < 0 -> add("El kilometraje no puede ser negativo")
                kmInicio != null && odometroKm < kmInicio ->
                    add("El kilometraje final ($odometroKm) no puede ser menor al de salida ($kmInicio)")
            }
            if (combustiblePct(combustible) == null) add("Elige el nivel de combustible")
        }
        return Validacion(faltantes = faltantes, errores = errores)
    }

    // ── `meta` multipart ────────────────────────────────────────────────────

    /**
     * El JSON de la parte `meta`: una entrada por casilla capturada, en el orden
     * de [SLOTS]. Las casillas sin hora no se incluyen (el servidor las rechazaría).
     */
    fun metaJson(metas: Map<String, SlotMeta>): String =
        SLOTS.mapNotNull { slot ->
            val meta = metas[slot.id]?.takeIf { it.valida } ?: return@mapNotNull null
            "${jsonString(slot.id)}:${slotJson(meta)}"
        }.joinToString(prefix = "{", separator = ",", postfix = "}")

    /** El JSON de una sola casilla (parte `meta-<slot>`, la otra forma que acepta el API). */
    fun slotJson(meta: SlotMeta): String {
        val campos = buildList {
            add("${jsonString("capturedAt")}:${jsonString(meta.capturedAt.orEmpty())}")
            meta.lat?.let { add("${jsonString("lat")}:$it") }
            meta.lng?.let { add("${jsonString("lng")}:$it") }
        }
        return campos.joinToString(prefix = "{", separator = ",", postfix = "}")
    }

    private fun jsonString(value: String): String {
        val sb = StringBuilder(value.length + 2)
        sb.append('"')
        for (c in value) {
            when (c) {
                '"' -> sb.append("\\\"")
                '\\' -> sb.append("\\\\")
                '\n' -> sb.append("\\n")
                '\r' -> sb.append("\\r")
                '\t' -> sb.append("\\t")
                else -> if (c < ' ') sb.append("\\u%04x".format(c.code)) else sb.append(c)
            }
        }
        sb.append('"')
        return sb.toString()
    }
}
